driver/src/main/com/mongodb/client/gridfs/GridFSBucket.java

```diff
@@ -0,0 +1,251 @@
+     *          application will write the contents.
+     */
+    GridFSUploadStream openUploadStream(String filename);
+
+    /**
+     * Opens a Stream that the application can write the contents of the file to.
+     *<p>
+     * As the application writes the contents to the returned Stream, the contents are uploaded as chunks in the chunks collection. When
+     * the application signals it is done writing the contents of the file by calling close on the returned Stream, a files collection
+     * document is created in the files collection.
+     *</p>
+     * @param filename the filename for the stream
+     * @param options the GridFSUploadOptions
+     * @return the GridFSUploadStream that provides the ObjectId for the file to be uploaded and the Stream to which the
+     *          application will write the contents.
+     */
+    GridFSUploadStream openUploadStream(String filename, GridFSUploadOptions options);
+
+    /**
+     * Uploads a user file to a GridFS bucket.
+     *<p>
+     * Reads the contents of the user file from the {@code Stream} and uploads it as chunks in the chunks collection. After all the
+     * chunks have been uploaded, it creates a files collection document for {@code filename} in the files collection.
+     *</p>
+     *
+     * @param filename the filename for the stream
+     * @param source the Stream providing the file data
+     * @return the ObjectId of the uploaded file.
+     */
+    ObjectId uploadFromStream(String filename, InputStream source);
+
+    /**
+     * Uploads a user file to a GridFS bucket.
+     * <p>
+     * Reads the contents of the user file from the {@code Stream} and uploads it as chunks in the chunks collection. After all the
+     * chunks have been uploaded, it creates a files collection document for {@code filename} in the files collection.
+     * </p>
+     *
+     * @param filename the filename for the stream
+     * @param source the Stream providing the file data
+     * @param options the GridFSUploadOptions
+     * @return the ObjectId of the uploaded file.
+     */
+    ObjectId uploadFromStream(String filename, InputStream source, GridFSUploadOptions options);
+
+    /**
+     * Opens a Stream from which the application can read the contents of the stored file specified by {@code id}.
+     *
+     * @param id the ObjectId of the file to be put into a stream.
+     * @return the stream
+     */
+    GridFSDownloadStream openDownloadStream(ObjectId id);
+
+    /**
+     * Downloads the contents of the stored file specified by {@code id} and writes the contents to the {@code destination} Stream.
+     *
+     * @param id the custom id of the file, to be written to the destination stream
+     * @param destination the destination stream
+     * @deprecated using custom id values for with GridFS is no longer supported
+     */
+    @Deprecated
+    void downloadLegacyFileToStream(Object id, OutputStream destination);
```

Comment:
Does the word "file" or "legacy" need to be in this method name?  How about:

    void downloadToStream(BsonValue id, OutputStream destination)

I don't mind making this a bit difficult by forcing the use of BsonValue in order to make the method name cleaner.